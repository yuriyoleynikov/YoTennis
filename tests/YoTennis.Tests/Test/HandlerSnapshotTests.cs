using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Text;
using Xunit;
using YoTennis.Models;
using YoTennis.Models.Events;

namespace YoTennis.Tests.Test
{
    class Scenario
    {
        private JsonSerializerSettings JsonSettings = new JsonSerializerSettings
        {
            Formatting = Formatting.Indented,
            TypeNameHandling = TypeNameHandling.Auto,
            Converters = { new Newtonsoft.Json.Converters.StringEnumConverter() }
        };
        private JsonSerializer Serializer = new JsonSerializer { Converters = { new Newtonsoft.Json.Converters.StringEnumConverter() } };

        private GameHandler _handler = new GameHandler();
        private StringBuilder _result = new StringBuilder();
        private DateTime _nextTime = new DateTime(1986, 09, 26);

        public Scenario()
        {
            _result.Append("Initial state ");
            DumpState();
            _result.AppendLine();
        }

        private void DumpState()
        {
            var state = JsonConvert.SerializeObject(_handler.CurrentState, JsonSettings);
            _result.Append(state);
        }

        private void DumpChange(JToken prevState, JToken nextState, string indent = "")
        {
            if (JToken.DeepEquals(prevState, nextState))
            {
                _result.Append("<no changes>");
            }
            else if (prevState.Type == JTokenType.Object && nextState.Type == JTokenType.Object)
            {
                DumpObjectDiff((JObject)prevState, (JObject)nextState, indent + "  ");
            }
            else
            {
                _result.Append($"{JsonConvert.SerializeObject(prevState, JsonSettings)} -> {JsonConvert.SerializeObject(nextState, JsonSettings)}");
            }
        }

        private void DumpObjectDiff(JObject prevObject, JObject nextObject, string indent = "")
        {
            _result.AppendLine("{");
            foreach (var prevProperty in prevObject)
            {
                var nextProperty = nextObject.Property(prevProperty.Key);
                if (!JToken.DeepEquals(prevProperty.Value, nextProperty.Value))
                {
                    _result.Append($"{indent}\"{prevProperty.Key}\": ");
                    DumpChange(prevProperty.Value, nextProperty.Value, indent);
                    _result.AppendLine(",");
                }
            }
            _result.Append("}");
        }

        public void AddEvent(GameEvent e)
        {
            if (e.OccuredAt == default(DateTime))
                e.OccuredAt = _nextTime;
            _nextTime = e.OccuredAt.AddMinutes(1);

            _result.AppendLine();
            _result.AppendLine($"Event {e.GetType().Name} {JsonConvert.SerializeObject(e, JsonSettings)}");
            var prevState = JToken.FromObject(_handler.CurrentState, Serializer);

            try
            {
                _handler.AddEvent(e);
            }
            catch (Exception ex)
            {
                _result.AppendLine($"Exception {ex.Message}");
            }

            var nextState = JToken.FromObject(_handler.CurrentState, Serializer);
            _result.Append("State ");
            DumpChange(prevState, nextState);
            _result.AppendLine();
        }

        public override string ToString() => _result.ToString();
    }

    public class HandlerSnapshotTests
    {
        [Fact]
        public void Test1()
        {
            var scenario = new Scenario();

            scenario.AddEvent(new StartEvent
            {
                FirstPlayer = "Oleynikov",
                SecondPlayer = "Nadal",
                Settings = new MatchSettings
                {
                    SetsForWin = 3,
                    TiebreakFinal = false
                }
            });
            scenario.AddEvent(new DrawEvent
            {
                PlayerOnLeft = Player.First,
                PlayerServes = Player.Second
            });
            scenario.AddEvent(new StartTiebreakEvent());

            Approvals.Verify(scenario.ToString());
        }
    }
}
