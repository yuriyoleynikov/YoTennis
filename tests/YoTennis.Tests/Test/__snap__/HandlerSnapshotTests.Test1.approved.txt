Initial state {
  "MatchScore": {
    "FirstPlayer": 0,
    "SecondPlayer": 0
  },
  "Sets": null,
  "GameScore": {
    "FirstPlayer": 0,
    "SecondPlayer": 0
  },
  "GameStratedAt": "0001-01-01T00:00:00",
  "PlayerOnLeft": "First",
  "PlayerServes": "First",
  "ServePosition": "Right",
  "SecondServe": false,
  "MatchStartedAt": "0001-01-01T00:00:00",
  "MatchSettings": null,
  "FirstPlayer": null,
  "SecondPlayer": null,
  "State": "NotStarted"
}

Event {
  "Settings": {
    "SetsForWin": 3,
    "TiebreakFinal": false,
    "GamesInSet": 6,
    "PointsInGame": 4,
    "PointsInTiebreak": 7
  },
  "FirstPlayer": "Oleynikov",
  "SecondPlayer": "Nadal",
  "OccuredAt": "1986-09-26T00:00:00"
}
State {
  "MatchStartedAt": "0001-01-01T00:00:00" -> "1986-09-26T00:00:00",
  "MatchSettings": null -> {
  "SetsForWin": 3,
  "TiebreakFinal": false,
  "GamesInSet": 6,
  "PointsInGame": 4,
  "PointsInTiebreak": 7
},
  "FirstPlayer": null -> "Oleynikov",
  "SecondPlayer": null -> "Nadal",
  "State": "NotStarted" -> "Drawing",
}

Event {
  "PlayerOnLeft": "First",
  "PlayerServes": "Second",
  "OccuredAt": "1986-09-26T00:01:00"
}
State {
  "Sets": null -> [
  {
    "Score": {
      "FirstPlayer": 0,
      "SecondPlayer": 0
    },
    "TiebreakScore": {
      "FirstPlayer": 0,
      "SecondPlayer": 0
    }
  }
],
  "GameStratedAt": "0001-01-01T00:00:00" -> "1986-09-26T00:01:00",
  "PlayerServes": "First" -> "Second",
  "State": "Drawing" -> "BeginningGame",
}

Event {
  "OccuredAt": "1986-09-26T00:02:00"
}
Exception Not Expected
State <no changes>
